### PR TITLE
Add integration test for inbound agent with ACI and simplify direct connection for inbound agents

### PR DESCRIPTION
**Pull request by timja**
* resolves conflicts in https://github.com/jenkinsci/azure-container-agents-plugin/pull/10 and makes it work properly
* update azure sdk, required to get the test to pass, was getting library conflicts
* simplify direct connection  by adding replacement for `${instanceIdentity}` https://github.com/jenkinsci/remoting/blob/cd622003d82ae80190d4a3c343dbad95ce076d1f/docs/inbound-agent.md#connect-directly-to-tcp-port, makes tunnelling easier for testing inbound agents
* some minor text / help cleanup
* change docs to pull from github instead of wiki
* use `RealJenkinsRule` to add a test proving classloading works
* Remove unneeded maskClasses

### CODE REVIEW

**Comment by jglick** [comment]
It _should_ “support” `maskClasses`, meaning that it should not be bypassing the Jenkins core class loader code that implements that option, so tests should fail iff the plugin would actually throw a `LinkageError` in production. Please verify though!

**Comment by jglick** [comment]
Obviously there are some trade-offs compared to ATH. The possible weak points:

* use of `*.jpl` rather than `*.jpi`
* stuff accessible to the test class loader somehow leaking “up” into plugin class loaders

**Comment by timja** [comment]
> It _should_ “support” `maskClasses`, meaning that it should not be bypassing the Jenkins core class loader code that implements that option, so tests should fail iff the plugin would actually throw a `LinkageError` in production. Please verify though!

I mean I have verified and doing it manually in production works but not via `RealJenkinsRule`

File: pom.xml [hunk 1]
```diff
@@ -284,12 +298,14 @@
                     <minimumJavaVersion>1.8</minimumJavaVersion>
                     <pluginFirstClassLoader>true</pluginFirstClassLoader>
                     <maskClasses>
```
Comment (jglick) [inline pom.xml:300]: Is there a reason this plugin is using _both_ of these systems? That seems surprising.

**Comment by jglick** [comment]
I am confused by what you are saying. You mean the usage of `maskClasses` seems to work in “production” (`jenkins.war` + `$JENKINS_HOME/plugins/azure-container-agents.jpi`) but gives an error from the `RealJenkinsRule` test? Yet all tests are passing here in CI; but I do not see the new test in https://ci.jenkins.io/job/Plugins/job/azure-container-agents-plugin/job/PR-66/2/testReport/com.microsoft.jenkins.containeragents/ ?

**Comment by timja** [comment]
The new test can’t run in CI as it needs azure credentials to run, if you run it via maven or ide with appropriate credentials it fails

yes it works in production but not via real Jenkins rule

File: pom.xml (hunk 1 above)
Comment (timja) [reply pom.xml:300]: Erm I’ll try remove plugin first class loader, it doesn’t work on its own without mask classes iirc

**Comment by jglick** [comment]
> if you run it via maven or ide with appropriate credentials

Not something I could do myself.

Not sure offhand. You would need to check that the defining loader for various relevant types is the expected one. Unfortunately Java 8 does not display meaningful `NoClassDefFoundError`’s (some update or another does better) so you have to manually insert debug code, or set breakpoints and inspect. For example, from `AciService.createDeployment` try something like

```java
ClassLoader cl = com.microsoft.rest.Validator.class.getClassLoader();
System.err.println(cl);
System.err.println(com.microsoft.rest.Validator.class.getProtectionDomain().getCodeSource().getLocation());
System.err.println(cl.loadClass("com.google.common.reflect.TypeToken"));
```

**Comment by timja** [comment]
K it works now

The guava error just seemed to disappear not sure what was up there,
I had to avoid using `TokenCredentialData` otherwise I got linkageerrors

I'll do some cleanup, write a quicker test to check class loading as this takes 5minutes =/

File: pom.xml (hunk 1 above)
Comment (timja) [reply pom.xml]: Wasn't able to remove `pluginFirstClassLoader` was getting weird conflicts with `okhttp`.

But removing `maskClasses` worked...

**Pull request merged**